### PR TITLE
Add an action to generate kernels for grand potential model

### LINKED ISSUE #11386: Add an action to generate kernels for Grand Potential Model
## Rationale

The grand potential model in the phase field module requires a large number of kernels. It will be much easier to use with an action that automatically generates those kernels for you.

## Description

- Action that generates kernels related to grand potential model
- New test of said action
- Modification of ACGrGrMulti Kernel to allow for specification of material property "mu" by name
- Modifications to FeatureFloodCount and FeatureVolumeVectorPostProcessor that allow for optional outputing of feature centroids in the vectorpostprocessor
- Supporting changes to PhaseFieldApp.C and tests

## Impact

New phase field module action object, minor modifications of existing objects, new test, easier to develop grand potential model input files

### PR DESCRIPTION
**Pull request by itgreenquist**
closes #11386

Adds a new action to the phase field module to generate grand potential model kernels. The grand potential model requires a high number of kernels and this should make it easier to make input files.

### CODE REVIEW

File: modules/phase_field/include/postprocessors/FeatureFloodCount.h [hunk 1]
```diff
@@ -62,6 +62,7 @@ class FeatureFloodCount : public GeneralPostprocessor,
 
   /// Returns a Boolean indicating whether this feature intersects _any_ boundary
   virtual bool doesFeatureIntersectBoundary(unsigned int feature_id) const;
+  virtual Point featureCentroid(unsigned int feature_id) const;
```
Comment (dschwen) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: This looks unrelated to the referenced issue.

File: modules/phase_field/src/action/GPMKernelAction.C [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+      getParam<std::vector<MaterialPropertyName>>("Fj_op");
+  const MaterialPropertyName kappa_gr = getParam<MaterialPropertyName>("kappa_gr");
+  const MaterialPropertyName kappa_op = getParam<MaterialPropertyName>("kappa_op");
+  const MaterialPropertyName gamma_gr = getParam<MaterialPropertyName>("gamma_gr");
+  const MaterialPropertyName gamma_op = getParam<MaterialPropertyName>("gamma_op");
+  const MaterialPropertyName gamma_xx = getParam<MaterialPropertyName>("gamma_grxop");
+  const MaterialPropertyName gr_mob = getParam<MaterialPropertyName>("gr_mob_name");
+  const MaterialPropertyName op_mob = getParam<MaterialPropertyName>("op_mob_name");
+  const MaterialPropertyName en_barr_gr = getParam<MaterialPropertyName>("en_barr_gr");
+  const MaterialPropertyName en_barr_op = getParam<MaterialPropertyName>("en_barr_op");
```
Comment (dschwen) [inline modules/phase_field/src/action/GPMKernelAction.C]: I strongly suggest using `auto`

File: modules/phase_field/src/base/PhaseFieldApp.C [hunk 3]
```diff
@@ -99,6 +99,7 @@ PhaseFieldApp::associateSyntax(Syntax & syntax, ActionFactory & action_factory)
   registerSyntax("PolycrystalVoronoiVoidICAction", "ICs/PolycrystalICs/PolycrystalVoronoiVoidIC");
   registerSyntax("RigidBodyMultiKernelAction", "Kernels/RigidBodyMultiKernel");
   registerSyntax("Tricrystal2CircleGrainsICAction", "ICs/PolycrystalICs/Tricrystal2CircleGrainsIC");
+  registerSyntax("GPMKernelAction", "Kernels/GPMKernels");
```
Comment (dschwen) [inline modules/phase_field/src/base/PhaseFieldApp.C]: I'd perefer the new syntax path convention 

Modules/PhaseField/GrandPotential

Let's avoid user-visible acronyms

File: modules/phase_field/src/vectorpostprocessors/FeatureVolumeVectorPostprocessor.C [hunk 4]
```diff
@@ -33,7 +33,7 @@ validParams<FeatureVolumeVectorPostprocessor>()
                         "Set this Boolean if you wish to use an element based volume where"
                         " the dominant order parameter determines the feature that accumulates the "
                         "entire element volume");
-
+  params.addParam<bool>("output_centroids", false, "Set to true to output the feature centroids");
```
Comment (dschwen) [inline modules/phase_field/src/vectorpostprocessors/FeatureVolumeVectorPostprocessor.C]: Please remove unrelated changes (but out of curiosity, have you thought about periodicity?)

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C [hunk 5]
```diff
@@ -632,6 +632,23 @@ FeatureFloodCount::doesFeatureIntersectBoundary(unsigned int feature_id) const
   return false;
 }
 
+Point
+FeatureFloodCount::featureCentroid(unsigned int feature_id) const
+{
+  Point p;
+  if (feature_id >= _feature_id_to_local_index.size())
+    return p;
```
Comment (permcody) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: I'm not sure I'd return a (0,0,0) where processors may be missing information. That point usually is in the domain so it's hard to tell a valid result from an invalid one. As you can see with similar getters, usually the information isn't complete so I have a lot of protection built in to prevent people from getting incorrect information when coupling. If you need me to help out with making sure that you can obtain Centroids in all cases, let me know. I'm not entirely sure if this code is completely safe in parallel or not without a little looking around.

**Comment by itgreenquist** [comment]
Where can I find the requirements for the documentation test? Are they listed anywhere?

**Comment by permcody** [comment]
Not exactly concise yet (we are working on that), but here's an overview. Basically, you have to document each *new* class added. Old ones are grandfathered in for now until we can go back and document everything.

http://mooseframework.org/moose/utilities/MooseDocs/standards.html

File: modules/phase_field/doc/content/documentation/systems/Modules/PhaseField/GrandPotential/index.md [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<!-- MOOSE Documentation Stub: Remove this when content is added. -->
```
Comment (dschwen) [inline modules/phase_field/doc/content/documentation/systems/Modules/PhaseField/GrandPotential/index.md]: @aeslaughter, what is the purpose of these empty index files? I suppose they are useful if there are multiple objects below a new syntax hierarchy. Is there any link from this page to the actual documentation meat? (I guess I can look myself..)

File: modules/phase_field/doc/content/documentation/systems/Modules/PhaseField/GrandPotential/index.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+
+
+# GrainGrowth System
```
Comment (dschwen) [inline modules/phase_field/doc/content/documentation/systems/Modules/PhaseField/GrandPotential/index.md]: This should be _GrandPotential System_

File: modules/phase_field/include/action/GPMKernelAction.h [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+#ifndef GPMKERNELACTION
+#define GPMKERNELACTION
+
+/**
+ * Generates the necessary kernels for the Grand Potential Function for any number
+ * of order parameters and chemical potentials. Can do so with two sets of order
+ * parameters that use different material properties. Also able to use anisotropic
+ * diffusivities for the chemical potential variables.
+ **/
```
Comment (dschwen) [inline modules/phase_field/include/action/GPMKernelAction.h]: could you please move the doxygen comment directly above the class declaration

File: modules/phase_field/include/action/GPMKernelAction.h [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef GPMKERNELACTION
+#define GPMKERNELACTION
```
Comment (dschwen) [inline modules/phase_field/include/action/GPMKernelAction.h]: `GPMKERNELACTION_H`

File: modules/phase_field/include/action/GPMKernelAction.h [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+{
+public:
+  GPMKernelAction(const InputParameters & parameters);
+
+  virtual void act();
+};
+
+template <>
+InputParameters validParams<GPMKernelAction>();
+#endif // GPMKERNELACTION
```
Comment (dschwen) [inline modules/phase_field/include/action/GPMKernelAction.h]: space above

File: modules/phase_field/src/action/GPMKernelAction.C [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+  -TimeDerivative
+  -ACInterface
+  -ACSwitching
+  -ACGrGrMulti
+additional_order_parameters
+  -TimeDerivative
+  -ACInterface
+  -ACSwitching
+  -ACGrGrMulti
+*/
```
Comment (dschwen) [inline modules/phase_field/src/action/GPMKernelAction.C]: this should go in the doxygen comment IMO

File: modules/phase_field/test/tests/actions/gpm_kernel.i [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+[]
+
+[Executioner]
+  type = Transient
+  scheme = bdf2
+  solve_type = NEWTON
+  #petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
+  #petsc_options_value = ' lu       mumps'
+  petsc_options_iname = '-pc_type -sub_pc_type -pc_asm_overlap -ksp_gmres_restart -sub_ksp_type'
+  petsc_options_value = ' asm      lu           1               31                 preonly'
```
Comment (dschwen) [inline modules/phase_field/test/tests/actions/gpm_kernel.i]: does this really need all these PETSc options? Please remove commented out stuff.

**Review by dschwen — COMMENTED** [review]
Over all I think this is a great contribution that will make the use of the Grand Potential model a lot more convenient. So thanks!

I have a bunch of menial comments that I added (not menial enough to just ignore). As a general remark I think that a lot of the parameter names are necessarily cryptic. I.e. the abbreviation `en_barr_mu` saves no typing work with autocompletion but is way less expressive than `energy_barrier_mu`. A local variable `gam` is confusing when it is spelled out as `gamma` in comparable variables, etc.

File: modules/phase_field/doc/content/documentation/systems/Modules/PhaseField/GrandPotential/index.md (hunk 6 above)
Comment (dschwen) [reply modules/phase_field/doc/content/documentation/systems/Modules/PhaseField/GrandPotential/index.md]: Ok, there is an automatically generated link, so this is fine.

File: modules/phase_field/include/action/GPMKernelAction.h [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Generates the necessary kernels for the Grand Potential Function for any number
+ * of order parameters and chemical potentials. Can do so with two sets of order
+ * parameters that use different material properties. Also able to use anisotropic
+ * diffusivities for the chemical potential variables.
+ **/
+
+#include "Action.h"
+
+class GPMKernelAction : public Action
```
Comment (dschwen) [inline modules/phase_field/include/action/GPMKernelAction.h]: The class should be called `GrandPotentialKernelAction`

**Review by dschwen — COMMENTED** [review]
One more comment

**Comment by brianmoose** [comment]
@itgreenquist If you accept the invitation to be a collaborator on the moose repo, testing will be started automatically for you without a developer manually activating the tests.

**Comment by itgreenquist** [comment]
@brianmoose I did not know that I received an invitation. How do I accept it?

File: modules/phase_field/test/tests/actions/gpm_kernel.i [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+  nl_max_its = 20
+  l_max_its = 30
+  l_tol = 1e-4
+  nl_rel_tol = 1e-7
+  nl_abs_tol = 1e-7
+  start_time = 0
+  dt = 2e-5
+  num_steps = 10
```
Comment (laagesen) [inline modules/phase_field/test/tests/actions/gpm_kernel.i]: 1 or 2 time steps should be enough, don't want to take more than is needed to keep the test suite running as fast as possible

**Review by laagesen — CHANGES_REQUESTED** [review]
@itgreenquist,  I've added a few comments. Thanks for the contribution, this will be very useful!

**Comment by brianmoose** [comment]
I think it should have sent an email. If you can't find that you should be able to go to
https://github.com/idaholab (while logged in) and accept it there.

File: modules/phase_field/src/kernels/ACGrGrMulti.C [hunk 15]
```diff
@@ -21,14 +21,16 @@ validParams<ACGrGrMulti>()
       "gamma_names",
       "List of gamma material property names for each other order parameter. Place "
       "in same order as order parameters (v)!");
+  params.addParam<MaterialPropertyName>("en_barr", "mu", "Energy Barrier Coefficient");
```
[the remark this answers is not in the document]
Comment (itgreenquist) [reply modules/phase_field/src/kernels/ACGrGrMulti.C]: In my model I found that interface widths are more consistent if a separate mu is defined for the grain boundaries and surfaces. Currently, the same mu is automatically used for all cases, so I added this input to let you use multiple mu's.

Once I did that though, the only way I could get the code to compile was to reinitialize it in the header. If there is a way to keep this input while initializing in `ACGrGrBase`, please let me know.

[the remark this answers is not in the document]
Comment (laagesen) [reply modules/phase_field/src/kernels/ACGrGrMulti.C]: Hm... this model is set up to use a consistent `mu` throughout. Are you saying you are using different values of `mu` in different kernels? If you do that, I'm afraid your evolution equations are no longer variationally derived from the free energy functional, so they do not guarantee decreasing energy.

Using large/small values of gamma changes interface width. Moelans addressed this in Acta Mater. 59, 1077-1086 (2011), and proposed to make interface widths more consistent by making kappa a function of order parameters. I am not a proponent of that approach because the evolution equations are again non-variational, but I think the governing equations in that model could be fixed by including the order parameter dependence in the kappas. Similarly, you could make mu a function of the order parameters and take the variational derivative of mu accounting for its kappa dependence.

Sorry I know this is opening a can of worms you probably didn't intend to with this PR, but unless we can implement this strategy I think we should keep mu to be a single value throughout currently.

[the remark this answers is not in the document]
Comment (laagesen) [reply modules/phase_field/src/kernels/ACGrGrMulti.C]: The interfacial energy in general is controlled by mu, kappa, and gamma- if you've seen the manuscript I wrote on the model, see eq. (43) (I call mu m there to be consistent with the 2011 Moelans paper - if you don't have the manuscript let me know and I can send it to you.) So to set interfacial energies, we keep mu and kappa constant throughout, and vary gamma. So we would want to change gamma to get different energies between different types of interfaces. Does that make sense?

File: modules/phase_field/include/kernels/ACGrGrMulti.h [hunk 16]
```diff
@@ -38,6 +38,7 @@ class ACGrGrMulti : public ACGrGrBase
 
   /// Values of gammas for each order parameter
   std::vector<const MaterialProperty<Real> *> _prop_gammas;
+  const MaterialProperty<Real> & _mu;
```
[the remark this answers is not in the document]
Comment (dschwen) [reply modules/phase_field/include/kernels/ACGrGrMulti.h]: Why is this still here. Larry is right. You are shadowing the base class member. This is not the right design.

File: modules/phase_field/src/kernels/ACGrGrMulti.C [hunk 17]
```diff
@@ -21,14 +21,16 @@ validParams<ACGrGrMulti>()
...
   return params;
 }
 
 ACGrGrMulti::ACGrGrMulti(const InputParameters & parameters)
   : ACGrGrBase(parameters),
     _gamma_names(getParam<std::vector<MaterialPropertyName>>("gamma_names")),
     _num_j(_gamma_names.size()),
-    _prop_gammas(_num_j)
+    _prop_gammas(_num_j),
+    _mu(getMaterialProperty<Real>("energy_barrier"))
```
Comment (dschwen) [inline modules/phase_field/src/kernels/ACGrGrMulti.C]: remove

File: modules/phase_field/src/kernels/ACGrGrMulti.C [hunk 18]
```diff
@@ -21,14 +21,16 @@ validParams<ACGrGrMulti>()
       "gamma_names",
       "List of gamma material property names for each other order parameter. Place "
       "in same order as order parameters (v)!");
+  params.addParam<MaterialPropertyName>("energy_barrier", "mu", "Energy Barrier Coefficient");
```
Comment (dschwen) [inline modules/phase_field/src/kernels/ACGrGrMulti.C]: remove

**Review by dschwen — CHANGES_REQUESTED** [review]

File: modules/phase_field/include/kernels/ACGrGrMulti.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (dschwen) [reply modules/phase_field/include/kernels/ACGrGrMulti.h]: I see now what you are doing (@laagesen pointed it out). You allow the user to specify a custom name. If the model allows for that the change should be made in the base class, not by shadowing a member. If the model does not allow for that, because it breaks the variational derivative it should not be done at all.

[the remark this answers is not in the document]
Comment (itgreenquist) [reply modules/phase_field/include/kernels/ACGrGrMulti.h]: After talking to Larry today, I'm taking this out and I will work on implementing this differently in the model.

[the remark this answers is not in the document]
Comment (laagesen) [reply modules/phase_field/include/kernels/ACGrGrMulti.h]: @itgreenquist if you want to do what we talked about on the phone earlier, you still need to have the same mu throughout, and instead you make mu a function of the order parameters. That can be done by making mu a DerivativeParsedMaterial or creating a new class for it. Also you will probably need a new kernel to account for the effect of the variational derivative on the (now order parameter dependent) mu.

**Comment by permcody** [comment]
Looks like this PR is ready for another review.

**Review by dschwen — APPROVED** [review]
